### ISSUE TITLE
Few questions about the ConceptVectors dataset and codelines

### ISSUE THREAD

**Issue by hwi0lee** [yihuaihong/ConceptVectors#3]
Hi, I have questions about the dataset and codes:

1. There are "QA" and "unrelated QA" keys in the ConceptVectors dataset, but only questions are included. How can it be unlearned and evaluated without the corresponding ground truth answers?

2. I can find "wikipedia_content" key as well, I wonder what they are for. I thought it might be used for the additional training, but I read from the previous question that you are using vanilla pretrained model so that's not it. 

3. I have question from your code (forget.py)

```    
inputs = tokenizer(QA + text_completion + unrelated_QA, return_tensors="pt", padding=True,return_token_type_ids=False).to('cuda')
n_new_tokens = 100
with torch.no_grad():
    generation_output = model.generate(  # mt.model
        **inputs,
        do_sample=False,
        max_new_tokens=100,
    )
outputs = tokenizer.batch_decode(generation_output[:, -n_new_tokens:], skip_special_tokens=True)
qa_answers = outputs[:len(QA)]
text_responses = outputs[len(QA):-len(unrelated_QA)]
unrelated_qa_answers = outputs[-len(unrelated_QA):]
assert len(qa_answers) == 10 and len(unrelated_qa_answers) == 50
```
    
I can see that you are feeding 3 different behavior-evaluating question at once, and retrieving each answers from the constant position of the output. I wonder how this is possible since answers have various lengths. Also, I'm curious why you passed 3 different questions at once, expecting models to answer every each of it separately. I thought it is also against the real-world use cases where users rarely query model multiple questions, whose contexts are quite distant, at single turn.

I must have missed some important details, so I would appreciate your explanation. Thank you for your time.

**Comment by yihuaihong** [comment]
Hi, Hwiyeong.

Thanks for your interest in our project.

1. We evaluate the unlearned model performance **by comparing its response with the vanilla model's response**, computing by the metric like BLEU or ROUGE-L score. If the score is higher, it means that there are consistency between the responses for a certain concept's questions, which means the impact of unlearning training is small. On the contrary(the score is lower), it shows that the impact of unlearning training is greater.

2. **"wikipedia_content" is the content that is used to do the Unlearning training for the vanilla model**. Then, the vanilla model becomes an unlearned model. Next, we compare the unlearned model obtained with the original vanilla model, to observe the unlearning impact. In the last issue, what I mean about the vanilla model, is that we directly perform unlearning on the vanilla model without extra modification for that.

3. **These all questions are processed by the model in parallel, and the final effect is the same as the result you pass into the model one by one.** They are separate and do not affect each other. We just use batch-inputting and batch-decoding to speed up its processing. 
       For the answers retrieval, during our testing, we found each response for a certain question will all exceed 100 tokens, so we use 100 tokens to truncate them to ensure that each answer is the same length("max_new_tokens=100"). But this is indeed a hidden danger and unreasonable if someone use tricks like temperature sampling, so some answers may be less than 100 tokens. Thanks for your great reminding here and we will revise our code! 

Please free to let me know if you have any other questions or unclear things :)

**Comment by hwi0lee** [comment]
Thanks for your detailed explanation, but there's still something unclear about my second question.

A. For the term "vanilla model", are you referring to the original checkpoint released by the model distributor, i.e., OLMO trained with Dolma dataset, or your customized checkpoint of which you additionally trained with "wikipedia_content" as a training corpus? If you intended the former, you just assume that there's enough wikipedia knowledge in the model since they were pretrained on the wikipedia dataset? 

B. What I understand from your answer is that it is the unlearning phase where "wikipedia_content" intervenes, so that the "wikipedia_content" is working as an unlearning ingredient, am I right? Then, is this needed only for the gradient based methods like GA variants, but not for your oracle method "Needle"?

**Comment by yihuaihong** [comment]
Thanks for your questions.

A. For the vanilla model, we refer to the original checkpoint released by the model distributor, like LLaMA2-7b-chat and OLMO-7b, we don't customize anything here. And it is known that the pretraining datasets of these two open-source models both contain the Wikipedia dataset. 
    One important point of our paper is that we would like to evaluate the unlearning on the **intrinsic/inherent** knowledge of LLM(introduced through pretraining), instead of the artificial knowledge(later introduced through extra ft).

B. The Wikipedia content is working as an unlearning ingredient when testing grad_ascent, grad_diff, dpo, npo, npo_KL, and Needle(we add the noise on certain vector initially, and then do the ft on this vector solely). In the knowledge editing methods, we use the wikidata triplets as the training inputs instead of the wikipedia_content because of there requirements.

**Comment by hwi0lee** [comment]
Thanks for your time and efforts! Everything's clear now. Also, congratulations on your acceptance to EMNLP, I'm really looking forward to the release!

**Comment by yihuaihong** [comment]
Thank you very much for your kind words! I sincerely appreciate your support😊.
